### PR TITLE
Added logloss measure to binary/multiclass tasks with prob output

### PR DESCRIPTION
**Pull request by catastrophic-failure**
Added a measure. See discussion and benchmark in #818.

### CODE REVIEW

**Comment by berndbischl** [comment]
1) thx
2) pls read your diff. you added an unrelated typo in line 12
3) please separate into 2 objects, measureLogloss and logloss, the latter calls the former, like for all other measures in measures.R
4) the note seems wrong an unrelated?
5) please add a handcrafted unit test in tests/testthat/test_base_measures
6) why did you remove the comment lines, which where in the code in the issue thread? i would also add a comment for the (hard to read) model.matrix call, what happens there.

**Comment by berndbischl** [comment]
link to kaggle page with explanation of the measure
https://www.kaggle.com/wiki/LogarithmicLoss

**Comment by catastrophic-failure** [comment]
> 2) pls read your diff. you added an unrelated typo in line 12
Oop, I'm really sorry, must have mistakenly clicked in that line and dragged the text
> 3) please separate into 2 objects, measureLogloss and logloss, the latter calls the former, like for all other measures in measures.R
Not all measures are set like that, that's why I thought it was ok. But I'm changing it.
> 4) the note seems wrong an unrelated?
Sorry, previous note, from the code I sent to Giuseppe, which was indeed an edit of `multiclass.auc`
> 5) please add a handcrafted unit test in tests/testthat/test_base_measures
Will do.
> 6) why did you remove the comment lines, which where in the code in the issue thread? i would also add a comment for the (hard to read) model.matrix call, what happens there.
The other measures had no comments, but I'll add them back.

**Comment by PhilippPro** [comment]
Style guide: https://github.com/rdatsci/PackagesInfo/wiki/R-Style-Guide
;)

File: R/measures.R [hunk 1]
```diff
@@ -344,6 +344,30 @@ multiclass.auc = makeMeasure(id = "multiclass.auc", minimize = FALSE, best = 1,
...
+  properties = c("classif", "classif.multi", "req.truth", "req.prob"),
+  name = "Logarithmic loss",
+  note = "Check its definition in https://www.kaggle.com/wiki/MultiClassLogLoss.",
+  fun = function(task, model, pred, feats, extra.args) {
+    measureLogloss(getPredictionProbabilities(pred), pred$data$truth)
+  }
+)
+
+measureLogloss = function(probabilities, truth){
+  eps = 1e-15;
```
Comment (jakob-r) [inline R/measures.R]: no `;` needed

File: R/measures.R [hunk 2]
```diff
@@ -344,6 +344,30 @@ multiclass.auc = makeMeasure(id = "multiclass.auc", minimize = FALSE, best = 1,
...
+  #let's confine the predicted probabilities to [eps,1-eps], so logLoss doesn't reach infinity under any circumstance
+  probabilities[probabilities > 1-eps] = 1-eps
+  probabilities[probabilities < eps] = eps
+  #We add this line because binary tasks only output one probability column
+  if(nlevels(truth) == 2L) probabilities = cbind(probabilities,1-probabilities)
```
Comment (jakob-r) [inline R/measures.R]: Now I will be *that* guy but we write 'if (' with a space

**Comment by larskotthoff** [comment]
Is this ready to be merged?

**Comment by catastrophic-failure** [comment]
@larskotthoff Well, it's still missing a thing I suppose: just now I perceived other measures have notes over them in `measures.R`, these are added by hand I suppose, right?

**Comment by larskotthoff** [comment]
Yes.

**Comment by larskotthoff** [comment]
In this branch, do a "git rebase master". This may give you some conflicts that you need to fix manually. When done, push to this branch.

**Comment by catastrophic-failure** [comment]
Hmm seems like I resolved the conflicts, but now there's a load of unnecessary(?) commits. Is that a problem? Should I have squashed them first? If anything's wrong I can of course restart this PR.

**Comment by berndbischl** [comment]
did you do a pull before your push? you need to push-force. i did that mistake a milion times.....

**Comment by berndbischl** [comment]
@catastrophic-failure 
I dont think you can repair this here. you need to open up a new PR. but i wanted to help you to figure out what you did wrong. after you rebased, did you try to push? and then git told you, you need to do a pull before? because that pull is wrong.

**Comment by catastrophic-failure** [comment]
I manually edited the conflict and forgot a reference line, it's visible in the last commit. Apart from that, I also merged the commits from master to my repo, and from my repo to this branch (looking now, this doesn't look necessary), but I suppose the build will work now.

**Comment by berndbischl** [comment]
yes, but i cannot merge these 20 commits and i dont think i can squash them at the end.
could you please copy this to a new PR if you cannot squash this on your side pls?

**Comment by catastrophic-failure** [comment]
@berndbischl just give me time to learn how to properly do it. I did a proper rebase now and used push-force. On my local end there's only one commit but for some reason there are 7 commits on remote. Still learning the ropes...

**Comment by catastrophic-failure** [comment]
@berndbischl looks like I'm learning how to do it :)

**Comment by berndbischl** [comment]
@catastrophic-failure 
Please note that your implementation might have been faster than the one from @giuseppec, but it was "less correct". there were quite a few corner cases where it was wrong. (order of column labels, not all labels present in truth factor,....)
Please note how I fixed this in master now.

**Comment by catastrophic-failure** [comment]
@berndbischl I see the commit. Now there's a new function exported to select the correct elements I'll also change #887. About the possibility of missing truth levels I never saw it happening:

```splus
    library(mlr)
    fit = train("classif.lda", subsetTask(iris.task, 1:100)) #there are no virginica samples
    pred = predict(fit, iris.task)
```

`pred$data$truth` still has all three levels.

**Comment by berndbischl** [comment]
> @berndbischl I see the commit. Now there's a new function exported to select the correct elements I'll also change #887.

ok thx. pls do so.

> About the possibility of missing truth levels I never saw it happening:

well this is (i think...) that we in mlr ENSURE that. but you have to note multiple things:
- predict can also be called with a data.frame
- the basic "measure" is exported. so people can use them in custom, more general situations. there we cant ensure anytrhing.
- why write less robust code, if the robust version is not so hard to create?

**Comment by catastrophic-failure** [comment]
@berndbischl Indeed, will try to keep in mind. I'm so rooted on my personal uses for mlr sometimes I forget there are several ways to use it.
